### PR TITLE
Seungheonoh/agora utils

### PR DESCRIPTION
**Pull request by SeungheonOh**
Check `CHANGELOG.md` for detail.

liqwid-labs/agora#147

### CODE REVIEW

**Comment by chfanghr** [comment]
IIRC https://github.com/Liqwid-Labs/liqwid-plutarch-extra/pull/39/commits/0ba9595f1a63d3c9eb56e9d2634a3a4ee17cb61c have been revert in one of the merge commits, so maybe your will have to make a revert-revert commit.

**Comment by SeungheonOh** [comment]
manually added the change logs because it was a merge that made the part go away, and I don't know how to fight with Git to revert that

**Comment by chfanghr** [comment]
Added a few more utils in https://github.com/Liqwid-Labs/agora/pull/149 and https://github.com/Liqwid-Labs/agora/pull/146 .

File: src/Plutarch/Extra/Maybe.hs [hunk 1]
```diff
@@ -151,3 +156,23 @@ pexpectJustC escape ma = tcont $ \f ->
     pmatch ma $ \case
         PJust v -> f v
         PNothing -> escape
+
+{- | Extract the value stored in a PMaybe container.
+     If there's no value, throw an error with the given message.
+     @since 1.3.0
+-}
+mustBePJust :: forall a s. Term s (PString :--> PMaybe a :--> a)
```
Comment (peter-mlabs) [inline src/Plutarch/Extra/Maybe.hs:202]: I think we could use better naming conventions with the other functions in this module (like `pfromDJust`.)

Comment (peter-mlabs) [reply src/Plutarch/Extra/Maybe.hs:202]: And likewise similar interfaces (passing in the error string vs not)

**Comment by SeungheonOh** [comment]
@chfanghr Added utility for liqwid-labs/agora#149. I don't think there's any addition for liqwid-labs/agora#146

File: src/Plutarch/Extra/MultiSig.hs [hunk 2]
```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- |
+Module     : Agora.MultiSig
```
Comment (peter-mlabs) [inline src/Plutarch/Extra/MultiSig.hs:10]: Module name needs an update

File: src/Plutarch/Extra/IsData.hs [hunk 3]
```diff
@@ -73,7 +75,7 @@ import Prelude
 newtype ProductIsData a = ProductIsData {unProductIsData :: a}
```
Comment (peter-mlabs) [inline src/Plutarch/Extra/IsData.hs:75]: Why do we need/want this? What is the point?

File: src/Plutarch/Extra/IsData.hs [hunk 4]
```diff
@@ -73,7 +75,7 @@ import Prelude
...
 
 -- | Variant of 'PConstantViaData' using the List repr from 'ProductIsData'
-newtype PConstantViaDataList (h :: Type) (p :: S -> Type) = PConstantViaDataList h
+newtype DerivePConstantViaDataList (h :: Type) (p :: S -> Type) = DerivePConstantViaDataList h
```
Comment (peter-mlabs) [inline src/Plutarch/Extra/IsData.hs:78]: Likewise here -- what is the point of this?

File: src/Plutarch/Extra/Monoid.hs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+          PIsData
+        , -- | @since 1.3.0
+          PEq
+        , -- | @since 1.3.0
+          PShow
+        )
+        via (DerivePNewtype PAll PBool)
+
+instance forall (s :: S). Semigroup (Term s PAll) where
+    (<>) = pcon . PAll .* (#&&) `on` punsafeCoerce
```
Comment (peter-mlabs) [inline src/Plutarch/Extra/Monoid.hs:34]: Do we really need/want `punsafeCoerce` here?

Comment (peter-mlabs) [reply src/Plutarch/Extra/Monoid.hs:34]: Likewise for below

**Review by peter-mlabs — CHANGES_REQUESTED** [review]
Looks good except for the comments.

I'm also still not familiar with the SOP/deriving stuff, but if it seems to work that I don't have a problem merging.

File: src/Plutarch/Extra/IsData.hs (hunk 3 above)
Comment (chfanghr) [reply src/Plutarch/Extra/IsData.hs:75]: For example: 

https://github.com/Liqwid-Labs/agora/blob/staging/agora/Agora/Proposal.hs#L319
https://github.com/Liqwid-Labs/agora/blob/staging/agora/Agora/Proposal.hs#L636

File: src/Plutarch/Extra/Maybe.hs (hunk 1 above)
Comment (chfanghr) [reply src/Plutarch/Extra/Maybe.hs:202]: The intention here is to show a error message. I am not sure if this should be upstreamed to be honest.

File: src/Plutarch/Extra/Monoid.hs (hunk 5 above)
Comment (chfanghr) [reply src/Plutarch/Extra/Monoid.hs:34]: It's fine to just use `punsafeCoerce` here, I suppose. Btw, using `pto` here will result in a type check error.

File: src/Plutarch/Extra/Maybe.hs (hunk 1 above)
Comment (SeungheonOh) [reply src/Plutarch/Extra/Maybe.hs]: I renamed it to `passertPJust` and `passertPDJust`

File: src/Plutarch/Extra/IsData.hs (hunk 4 above)
Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs:78]: Not sure. It's from Connor's PR.

**Comment by peter-mlabs** [comment]
Okay, I think this is good. I'll merge once CI passes

**Comment by SeungheonOh** [comment]
@peter-mlabs can this me merged?

File: src/Plutarch/Api/V1/ScriptContext.hs [hunk 6]
```diff
@@ -29,7 +34,7 @@ import Plutarch.Api.V1 (
     PCredential (..),
     PDatum,
     PDatumHash,
-    PMaybeData,
+    PMaybeData (..),
```
Comment (kozross) [inline src/Plutarch/Api/V1/ScriptContext.hs:37]: Don't wildcard import constructors: spell out the ones you're using.

File: src/Plutarch/Api/V1/ScriptContext.hs [hunk 7]
```diff
@@ -248,3 +253,64 @@ paddressFromPubKeyHash = plam $ \pkh stakingCred ->
...
+        ( PBuiltinList (PAsData (PTuple PDatumHash PDatum))
+            :--> PTxOut
+            :--> PMaybe PDatum
+        )
+pfindTxOutDatum = phoistAcyclic $
+    plam $ \datums out -> unTermCont $ do
+        datumHash' <- pmatchC $ pfromData $ pfield @"datumHash" # out
+        pure $ case datumHash' of
+            PDJust ((pfield @"_0" #) -> datumHash) -> pfindDatum # datumHash # datums
+            _ -> pcon PNothing
```
Comment (kozross) [inline src/Plutarch/Api/V1/ScriptContext.hs:316]: Missing `forall (s :: S) .` throughout.

File: src/Plutarch/Extra/Field.hs [hunk 8]
```diff
@@ -24,7 +24,7 @@ type family BindAll (ps :: [PLabeledType]) :: [Symbol] where
 
 {- | Same as `pletFields` but instead of specifiying fields, it will take all fields.
 
- @since 1.2.1
+ @since 1.3.0
 -}
 pletAll ::
     forall a s b ps bs.
```
Comment (kozross) [inline src/Plutarch/Extra/Field.hs:30]: Retrospective, but these all need kind signatures.

File: src/Plutarch/Extra/Field.hs [hunk 9]
```diff
@@ -40,7 +40,7 @@ pletAll = pletFields @(BindAll ps)
 
 {- | TermCont version of `pletAll`
 
- @since 1.2.1
+ @since 1.3.0
 -}
 pletAllC ::
     forall a s b ps bs.
```
Comment (kozross) [inline src/Plutarch/Extra/Field.hs:46]: Same here.

File: src/Plutarch/Extra/Function.hs [hunk 10]
```diff
@@ -2,10 +2,12 @@ module Plutarch.Extra.Function (
     pconst,
     pidentity,
     papply,
+    pon,
+    pbuiltinUncurry,
 ) where
 
 import Data.Kind (Type)
-import Plutarch (S, Term, phoistAcyclic, plam, (#), type (:-->))
+import Plutarch.Prelude (PAsData, PBuiltinPair, PIsData, S, Term, pfromData, pfstBuiltin, phoistAcyclic, plam, psndBuiltin, (#), type (:-->))
```
Comment (kozross) [inline src/Plutarch/Extra/Function.hs:10]: Break this into multiple lines.

File: src/Plutarch/Extra/IsData.hs [hunk 11]
```diff
@@ -23,6 +23,8 @@ module Plutarch.Extra.IsData (
 --------------------------------------------------------------------------------
 
 import Data.Coerce (coerce)
+import Data.Functor.Identity (Identity (..))
+import Data.Maybe (fromJust)
 import Data.Proxy (Proxy (..))
```
Comment (kozross) [inline src/Plutarch/Extra/IsData.hs:28]: Again somewhat retroactive: no wildcard imports.

File: src/Plutarch/Extra/IsData.hs [hunk 12]
```diff
@@ -152,6 +178,10 @@ instance (Enum a) => ToData (EnumIsData a) where
 instance (Enum a) => FromData (EnumIsData a) where
     fromBuiltinData = coerce $ fmap (toEnum @a . fromInteger) . fromBuiltinData @Integer
 
+-- | @since 1.1.0
+instance (Enum a) => UnsafeFromData (EnumIsData a) where
+    unsafeFromBuiltinData = coerce . toEnum @a . fromInteger . unsafeFromBuiltinData @Integer
+
 -- | @since 1.1.0
 instance
     ( forall s. Enum (a s)
```
Comment (kozross) [inline src/Plutarch/Extra/IsData.hs:187]: Somewhat retrospective, but we need a kind signature on `s` here.

File: src/Plutarch/Extra/List.hs [hunk 13]
```diff
@@ -18,13 +18,16 @@ module Plutarch.Extra.List (
...
+    pisUniqBy',
+    pisUniq',
     module Extra,
 ) where
 
 import Data.Kind (Type)
 import Plutarch (
     PCon (pcon),
     PMatch (pmatch),
+    PType,
```
Comment (kozross) [inline src/Plutarch/Extra/List.hs:30]: I thought we were removing uses of this?

File: src/Plutarch/Extra/Map.hs [hunk 14]
```diff
@@ -10,12 +11,15 @@ module Plutarch.Extra.Map (
...
     pkvPairLt,
+    pfoldlWithKey,
+    pfoldMapWithKey,
 ) where
 
 import Data.Foldable (foldl')
 import Data.Kind (Type)
 import Plutarch (
     PMatch (pmatch),
+    PType,
```
Comment (kozross) [inline src/Plutarch/Extra/Map.hs:22]: Same again.

File: src/Plutarch/Extra/Maybe.hs [hunk 15]
```diff
@@ -151,3 +194,23 @@ pexpectJustC escape ma = tcont $ \f ->
...
+
+{- | Extract the value stored in a PMaybeData container.
+     If there's no value, throw an error with the given message.
+     @since 1.3.0
+-}
+passertPDJust :: forall a s. (PIsData a) => Term s (PString :--> PMaybeData a :--> a)
+passertPDJust = phoistAcyclic $
+    plam $ \emsg mv' -> pmatch mv' $ \case
+        PDJust ((pfield @"_0" #) -> v) -> v
+        _ -> ptraceError emsg
```
Comment (kozross) [inline src/Plutarch/Extra/Maybe.hs:216]: Kind signatures missing.

File: src/Plutarch/Extra/Monoid.hs (hunk 5 above)
Comment (kozross) [reply src/Plutarch/Extra/Monoid.hs:34]: I don't think this is too bad here, but I feel it'd probably be better to `pmatch` then rewrap. Newtypes are invisible at runtime.

File: src/Plutarch/Extra/MultiSig.hs [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+module Plutarch.Extra.MultiSig (
+    validatedByMultisig,
+    pvalidatedByMultisig,
+    PMultiSig (..),
+    MultiSig (..),
+) where
+
+import qualified GHC.Generics as GHC (Generic)
+import Generics.SOP (Generic, I (I))
+import Plutarch.Api.V1 (PPubKeyHash, PTxInfo (..))
```
Comment (kozross) [inline src/Plutarch/Extra/MultiSig.hs:25]: No wildcard imports.

File: src/Plutarch/Extra/MultiSig.hs [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+            let signatories = pfield @"signatories" # txInfo
+            pure $
+                pfromData multi.minSigs
+                    #<= ( plength #$ pfilter
+                            # plam
+                                ( \a ->
+                                    pelem # a # pfromData signatories
+                                )
+                            # multi.keys
+                        )
```
Comment (kozross) [inline src/Plutarch/Extra/MultiSig.hs:160]: Kind signature for `s` missing.

File: src/Plutarch/Extra/Other.hs [hunk 18]
```diff
@@ -22,10 +25,12 @@ import Plutarch (
...
  )
 import Plutarch.Bool (PEq, POrd)
-import Plutarch.Builtin (PIsData)
+import Plutarch.Builtin (PAsData, PData, PIsData)
 import Plutarch.DataRepr (PDataFields (..))
 import Plutarch.Integer (PIntegral)
-import Plutarch.Internal (S (SI))
+import Plutarch.Internal (S (SI), punsafeCoerce)
+import Plutarch.Lift (DerivePConstantViaNewtype, PConstantDecl, PLift, PUnsafeLiftDecl (PLifted))
+import Plutarch.TryFrom (PTryFrom (..))
```
Comment (kozross) [inline src/Plutarch/Extra/Other.hs:33]: Slightly retrospective, but no wildcard imports please.

**Review by kozross — CHANGES_REQUESTED** [review]

**Comment by SeungheonOh** [comment]
This is not dependent on #43. We need to hold merge until we are done with it.

**Comment by peter-mlabs** [comment]
Should be good to go, but @chfanghr wants to check `DerivePDataFieldsViaDataList` a bit more before merging to `main`. Give the thumbs up when we're good!

**Review by peter-mlabs — APPROVED** [review]
Got the thumbs up, so this is getting merged :partying_face:

**Pull request merged**